[DOCS] Convert 'Restore a snapshot' to tutorial

jrodewig opened the pull request:
Updates the 'Restore a snapshot' guide to be more tutorial-focused.
Adds a tutorial for restoring an entire cluster.

Closes #72497.


### Previews
* Restore a snapshot: https://elasticsearch_76929.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/snapshots-restore-snapshot.html
* Restore snapshot API: https://elasticsearch_76929.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/restore-snapshot-api.html
* Monitor snapshot progress: https://elasticsearch_76929.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/snapshots-monitor-snapshot-restore.html

jrodewig committed:
[DOCS] Convert 'Restore a snapshot' to tutorial

Updates the `Restore a snapshot` guide to be more tutorial-focused. The current
page duplicates reference content from the restore snapshot API docs.

I plan to add a tutorial for restoring a full cluster in a later PR.

jrodewig committed:
Edit I forgot to commit

jrodewig committed:
Better instructs for unassigned shards

jrodewig committed:
Clarify

jrodewig committed:
Clarify

elasticmachine committed:
Merge branch 'master' into docs__restore-snapshot-tut

jrodewig committed:
Note about existing data streams

jrodewig committed:
Fix typo

jrodewig committed:
Better get snapshot query params

jrodewig committed:
more realistic snapshot name

stefnestor commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,175 +1,268 @@
+[[close-restore]]
+==== Close and restore
+
+You can restore an existing index if it's closed. However, the index in the
```
I believe this maybe is suppose to merge restore data if the index is only closed & not deleted. (aka. restores un-findable records) -- as a comment on why you'd choose one & not the other

jrodewig replied:
Unfortunately, this isn't the case. Restoring a closed index overwrites its existing data. It's more or less the same as delete and restore. I updated the copy to note this in 986850c69e5c9bf7a2534ffccd25906d38401276.

stefnestor commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,175 +1,268 @@
+[[rename-on-restore]]
+==== Rename on restore
+
+You can rename indices and data streams you restore. Before you start, ensure
```
This is frequently used when you want to side-by-side compare old vs current index data (e.g. unexpected record updates/deletes).

jrodewig replied:
I struggled with the wording, but I added a statement to cover this.

stefnestor commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,175 +1,268 @@
+// TEST[setup:setup-snapshots]
+// TEST[s/_restore/_restore?wait_for_completion=true/]
+
+When the restore operation is complete, you can delete the original indices and
```
When the restore operation is complete, **if you choose to overwrite the current index with the restored index**, you can delete the original indices and data streams ... and reindex the restored ...

jrodewig replied:
I reworded this a bit to make it clear that the delete + reindex is optional.

stefnestor commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,175 +1,268 @@
+==== Delete and restore
+
+The simplest way to avoid conflicts is to delete an existing index or data
+stream before restoring it. To prevent the accidental recreation of the index or
```
`before [restoring](https://www.elastic.co/guide/en/elasticsearch/reference/current/restore-snapshot-api.html) it`

jrodewig replied:
I link to the API above. The Kibana feature is also a valid method.

stefnestor commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,175 +1,268 @@
+[[cancel-restore]]
+=== Cancel a restore
+
+You can delete an index or data stream to cancel its ongoing restore.
```
`delete a __restoring__ index`

AFAICT this only makes sense for the restore & replace option. Because a deleted index, deleted again wouldn't end up with the original index. A closed index restore would delete the current data which would be unexpected... 🤔

jrodewig replied:
This isn't the case. Deleting an index or data stream also deletes its existing data, even if its a closed index. I added a statement to clarify this.

stefnestor commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,175 +1,268 @@
 [discrete]
-=== Restoring to a different cluster
+[[restore-different-cluster]]
+=== Restore to a different cluster
```
1. Maybe add disclaimer that ESS handles this for you & link https://www.elastic.co/guide/en/cloud/current/ec-restoring-snapshots.html#ec-restore-across-clusters
2. link version compatibility chart when talking about compatible restores: https://www.elastic.co/guide/en/elasticsearch/reference/current/snapshot-restore.html#snapshot-restore-version-compatibility

jrodewig replied:
I added xrefs for these.

stefnestor commented:
@jrodewig thoughts on these final points?

1. system indices
    > Wildcard warning for 7.x versions

    If users are migrating on-prem to Cloud, we do sometimes recommend full system restores, w/ESS walk through (ES would be the same via manual filebased user auth)

    ![image](https://user-images.githubusercontent.com/26751266/131396592-6e4acd29-7cb7-47a1-ab95-332b58612283.png)

2. _very_ common errors, esp. via default path on the Kibana UI

    ![image](https://user-images.githubusercontent.com/26751266/131396704-b1dcedb1-17a0-4e72-bde1-b435b0670873.png)

3. Enterprise Search databases need a disclaimer to not restore until you've inputted the database encryption key, otherwise your data restore won't unencrypt. Only seems to effect EntSearch

jrodewig committed:
Address review feedback

jrodewig committed:
Edits

jrodewig committed:
reword

jrodewig commented:
Thanks for the review @stefnestor. Apologies for the delayed response. I was on vacation for a bit. I have responses to your questions below:

> 1. What about a full cluster restore?

I'm handling that as a separate PR. That'll involve updating some existing documentation under Back up a cluster. We also need to include a step for restoring the cluster configuration.

> 2. What about common errors?

I added a troubleshooting section with 986850c69e5c9bf7a2534ffccd25906d38401276 that covers these errors.

> 3. Enterprise Search databases need a disclaimer to not restore until you've inputted the database encryption key, otherwise your data restore won't unencrypt. Only seems to effect EntSearch

It looks like we should at least link to [Configure encryption keys](https://www.elastic.co/guide/en/enterprise-search/current/encryption-keys.html). However, I'm not very familiar with Enterprise Search.

@chriscressman Do you mind providing some guidance? Maybe we can add a section to the page I linked above and link to that page from here? Happy to collaborate on other approaches too.

jrodewig committed:
reword

jrodewig committed:
reorder

jrodewig committed:
reword

jrodewig committed:
edits

jrodewig committed:
edits

jrodewig committed:
fix typos

jrodewig committed:
edit

jrodewig committed:
edit

chriscressman commented:
> @chriscressman Do you mind providing some guidance? Maybe we can add a section to the page I linked above and link to that page from here (perhaps as a prerequisite)? Happy to collaborate on other approaches too.

Hey James. Our encryption keys doc needs some improvement. I create [an issue](https://github.com/elastic/enterprise-search-team/issues/811) to track that work.

However, to unblock you, I think it would be sufficient to say this somewhere within your doc:

> If your snapshot contains data from App Search or Workplace Search, ensure you've restored the [Enterprise Search encryption key](https://www.elastic.co/guide/en/enterprise-search/current/encryption-keys.html) _before_ restoring the snapshot.

I hope that helps.

jrodewig commented:
Perfect. Thanks @chriscressman!

jrodewig committed:
Add prereq for enterprise search

Leaf-Lin commented:
Is it worth adding https://github.com/elastic/elasticsearch/issues/76153 as a current limitation on restoring data stream?

jrodewig commented:
Thanks @Leaf-Lin!

> Is it worth adding #76153 as a current limitation on restoring data stream?

I note that restoring an individual backing index doesn't add it to a data stream under the `indices` parameter in the
restore snapshot API docs:

```
You can use this option to restore only a specific backing index from a data
stream. However, the restore operation doesn't add the restored backing index to
any existing data stream.
```

While I don't feel strongly, it seemed most appropriate there for a couple of reasons:

1. I don't think you can restore individual backing indices using Kibana's restore feature. At least, I wasn't able to in my testing.

2. I didn't think it was it a super common use case. However, I may be wrong there.

If you think differently, let me know. I can add it to the tutorial if needed.

jrodewig committed:
Add tip for excluding system indices from restore request

jrodewig committed:
Fix wrap + missing period

jrodewig committed:
Fix snippet tests

jrodewig committed:
Minor fix

jrodewig committed:
Clarify

jrodewig committed:
Rewrite

jrodewig committed:
Cleanup

jrodewig committed:
add redirect

jrodewig committed:
fix whitespace

jrodewig committed:
format fix

jrodewig committed:
Merge remote-tracking branch 'upstream/master' into docs__restore-snapshot-tut

jrodewig committed:
fixes

jrodewig commented:
I made some additional changes to the PR:

* Added a **Considerations** section for limitations and gotchas.
* Removed the **Close and restore** method. This is adequately covered elsewhere now.
* Added a procedure for restoring an entire cluster.

@Leaf-Lin @DaveCTurner This is ready for a review at your convenience. Thanks!

jrodewig committed:
whitespace

jrodewig committed:
Clarify config restoration. Remove dupe config restore docs

jrodewig committed:
More config restore cleanup

DaveCTurner commented on docs/reference/snapshot-restore/apis/restore-snapshot-api.asciidoc:
```diff
@@ -118,14 +115,17 @@ operation initializes. Defaults to `false`.
 `ignore_index_settings`::
-(Optional, string)
-A comma-separated list of index settings that should not be restored from a snapshot.
+(Optional, array of string) Index settings to not restore from the snapshot. You
```
This can be either an array of strings or a single string containing the setting names separated by commas:

https://github.com/elastic/elasticsearch/blob/a5d2251064d3457dc253503ba9ce75dd57fa6f4e/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/restore/RestoreSnapshotRequest.java#L526-L530

jrodewig replied:
Good catch. Thanks!

DaveCTurner commented on docs/reference/snapshot-restore/apis/restore-snapshot-api.asciidoc:
```diff
@@ -171,23 +171,17 @@ By default, all available feature states will be restored if `include_global_sta
-+
-By default, a restore operation includes all data streams and indices in the snapshot. If this
-argument is provided, the restore operation only includes the specified data streams and indices.
+(Optional, string) Comma-separated list of indices and data streams to restore.
```
Likewise, this can be either:

https://github.com/elastic/elasticsearch/blob/a5d2251064d3457dc253503ba9ce75dd57fa6f4e/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/restore/RestoreSnapshotRequest.java#L489-L493

DaveCTurner commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,187 +1,523 @@
 == Restore a snapshot
 
-////
+This guide shows you how to restore a snapshot. Snapshots are a convenient way
+to transfer data between clusters. You can also restore a snapshot to recover
+indices and data streams after deletion or a hardware failure.
```
> Snapshots are a convenient way to transfer data between clusters.

Saying this first suggests that transferring data is their primary purpose but really they're more often used for recovery. How about something like this?

```suggestion
This guide shows you how to restore a snapshot. Snapshots are a convenient way
to store a copy of your data outside of a cluster. You can restore a snapshot
to recover indices and data streams after deletion or a hardware failure. You
can also use snapshots to transfer data between clusters.
```

DaveCTurner commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,187 +1,523 @@
+
+[discrete]
+[[retrieve-snapshot]]
+=== Retrieve a snapshot
```
"Retrieve" isn't quite the right word IMO, it suggests you're retrieving all the data the snapshot contains but really you're just getting a bit of information about its contents.

DaveCTurner commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,187 +1,523 @@
+To view a list of available snapshots in {kib}, go to the main menu and click
+*Stack Management > Snapshot and Restore*.
+
+You can also use the <<get-snapshot-repo-api,get repository API>> to fetch a
+list of registered snapshot repositories.
```
Suggest starting with an overview sentence.

```suggestion
You can also use the <<get-snapshot-repo-api,get repository API>> and the
<<get-snapshot-api,get snapshot API>> to find snapshots that are available to
restore. First use the <<get-snapshot-repo-api,get repository API>> to fetch a
list of registered snapshot repositories.
```

DaveCTurner commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,187 +1,523 @@
-to override index settings during the restore process. For example, the
-following request will restore the index `index_1` without creating any
-replicas while switching back to the default refresh interval:
+If you take dedicated snapshots of the `.security` system index, restore it as well.
```
Is this a thing people do? Do we recommend doing it elsewhere? If not I'd suggest leaving out this step. It's certainly true (and hopefully obvious) that if you don't have a full snapshot of the whole cluster then you'll have to restore it in pieces, but it's not the common case and I think it'll cause confusion to mention it here.

jrodewig replied:
We do recommend this here:

https://www.elastic.co/guide/en/elasticsearch/reference/master/security-backup.html#backup-security-index-configuration

I also confirmed with the @elastic/es-security team that we do still recommend dedicated security index backups in some cases.

However, I think you're right. if you do dedicated backups, it's clear that you need to restore those indices separately.
I'll remove this for now.

DaveCTurner commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,187 +1,523 @@
-The information stored in a snapshot is not tied to a particular cluster or a cluster name. Therefore, it's possible to
-restore a snapshot created from one cluster into another cluster by registering the repository that contains the snapshot in the new cluster and starting the restore process.
+You can delete an index or data stream to cancel its ongoing restore. This also
+deletes any existing data for the index or data stream.
```
Suggest clarifying somehow that we mean the existing data _in the cluster_, the data in the snapshot is unaffected.

DaveCTurner commented on docs/reference/snapshot-restore/restore-snapshot.asciidoc:
```diff
@@ -1,187 +1,523 @@
-* Use the `indices` parameter to choose only specific data streams or indices to restore.
+[discrete]
+[[restore-different-cluster]]
+=== Restore to a different cluster
```
I think it would help to add that you have to register the repository with the new cluster, and mention that it should have `"readonly": true` if it's registered with another cluster too.

DaveCTurner reviewed (commented):
Looks good @jrodewig, I left a few small comments.

jrodewig committed:
Address review feedback

jrodewig committed:
Remove security index restore

jrodewig commented:
Thanks for the feedback @DaveCTurner. I incorporated all your suggestions.

jrodewig committed:
Fix docs build

jrodewig committed:
Fix typo. Remove dupe sentence.

jrodewig committed:
Clarify cancel effect on snapshot

jrodewig commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into docs__restore-snapshot-tut

jrodewig commented:
@elasticmachine run elasticsearch-ci/docs-check

jrodewig commented:
Going to merge this in. Thanks for the reviews and feedback, everyone!

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  7.x  | Commit could not be cherrypicked due to conflicts |
| ❌ |  7.15  | Commit could not be cherrypicked due to conflicts |
| ❌ |  7.14  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 76929`

jrodewig commented:
Turns out Kibana links to https://www.elastic.co/guide/en/elasticsearch/reference/7.14/snapshots-restore-snapshot.html#change-index-settings-during-restore, which was removed with this PR.

I've opened https://github.com/elastic/kibana/pull/112597 to update the link in Kibana. However, that means I'll only backport this change to `7.x` (currently 7.16) and `master` (8.0), not 7.15 or 7.14.
